File: src/bearings/web/models/analytics.py
# mypy: disable-error-code=explicit-any
"""Pydantic wire shapes for ``routes/analytics.py``.

Per ``docs/architecture-v1.md`` §1.1.5 the wire DTOs live alongside the
route module.  These shapes mirror the five :mod:`bearings.db.analytics`
dataclasses for the read path, and define request bodies for the logging
path (``POST /api/analytics/turns`` and
``POST /api/analytics/plug-blocks/batch``).

Phase 1 ships the schema and these models; the route module that binds
them arrives in a later phase.  All field names follow the JSON
conventions from ``BEARINGS_ANALYTICS_v1.md`` §9.

The ``mypy: disable-error-code=explicit-any`` pragma matches the narrow
carve-out other ``web/models/*`` modules make for Pydantic's
metaclass-exposed ``Any`` surface.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from bearings.config.constants import (
    KNOWN_ANALYTICS_BLOCK_TYPES,
    KNOWN_ANALYTICS_WARNING_TYPES,
    PLUG_RED_THRESHOLD_TOKENS,
    PLUG_YELLOW_THRESHOLD_TOKENS,
)

# ---------------------------------------------------------------------------
# Logging endpoint request shapes (§9.1)
# ---------------------------------------------------------------------------

_BLOCK_TYPE_VALUES: list[str] = sorted(KNOWN_ANALYTICS_BLOCK_TYPES)
_WARNING_TYPE_VALUES: list[str] = sorted(KNOWN_ANALYTICS_WARNING_TYPES)


class TurnIn(BaseModel):
    """Request body for ``POST /api/analytics/turns`` (§9.1).

    Records one Claude API turn's token consumption.  All token fields
    default to 0 so callers can omit fields absent from the SDK response
    (e.g. cache tokens on non-caching models).

    ``model`` is mandatory and must be non-empty — every token count row
    carries the model so aggregation queries can group by model before
    summing (spec §3.2 tokenizer constraint).
    """

    model_config = ConfigDict(extra="forbid")

    session_id: str = Field(min_length=1)
    turn_index: int = Field(ge=0)
    model: str = Field(min_length=1)
    input_tokens: int = Field(ge=0, default=0)
    output_tokens: int = Field(ge=0, default=0)
    cache_read_tokens: int = Field(ge=0, default=0)
    cache_creation_tokens: int = Field(ge=0, default=0)


class PlugBlockIn(BaseModel):
    """One plug block descriptor inside a ``POST /api/analytics/plug-blocks/batch`` body."""

    model_config = ConfigDict(extra="forbid")

    hash: str = Field(min_length=64, max_length=64, description="sha256 hex digest")
    block_type: str = Field(description=f"One of: {_BLOCK_TYPE_VALUES}")
    content: str = Field(min_length=1)
    source_path: str | None = None


class PlugBlocksBatchIn(BaseModel):
    """Request body for ``POST /api/analytics/plug-blocks/batch`` (§9.1).

    Records the full set of plug blocks injected into a session at
    creation time.  ``model`` is the session's configured executor model,
    used to token-count each block on first insert (spec §5.3).
    """

    model_config = ConfigDict(extra="forbid")

    session_id: str = Field(min_length=1)
    model: str = Field(min_length=1)
    blocks: list[PlugBlockIn] = Field(min_length=1)


# ---------------------------------------------------------------------------
# Read endpoint response shapes (§9.2)
# ---------------------------------------------------------------------------


class BucketWindowOut(BaseModel):
    """One usage window (5-hour or weekly) inside :class:`BucketCurrentOut`."""

    model_config = ConfigDict(extra="forbid")

    used: int
    limit: int
    percent: float = Field(ge=0.0, le=100.0)


class BucketCurrentOut(BaseModel):
    """Response for ``GET /api/analytics/bucket/current`` (§9.2)."""

    model_config = ConfigDict(extra="forbid")

    five_hour: BucketWindowOut | None = None
    weekly: BucketWindowOut | None = None
    as_of: int = Field(description="unix ms timestamp of the snapshot")


class TagAttributionOut(BaseModel):
    """One row of ``GET /api/analytics/attribution`` (§9.2).

    ``tokens_by_model`` is a dict mapping model name → total tokens for
    that model within the requested window.  Callers must NOT sum across
    model keys without normalisation — the spec's §3.2 constraint
    (different tokenizers) means cross-model sums are misleading.
    ``share_total`` is the fraction of all tokens in the window consumed
    by this tag's sessions, computed after grouping by model.
    """

    model_config = ConfigDict(extra="forbid")

    tag: str
    tokens_by_model: dict[str, int]
    share_total: float = Field(ge=0.0, le=1.0)
    burn_rate_per_min: float = Field(ge=0.0)


class RedundancySessionRef(BaseModel):
    """Session reference inside a :class:`RedundancyBlockOut` row."""

    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    timestamp: int
    tags: list[str]


class RedundancyBlockOut(BaseModel):
    """One row of ``GET /api/analytics/redundancy`` (§9.2)."""

    model_config = ConfigDict(extra="forbid")

    hash: str
    block_type: str
    token_count: int
    token_count_model: str
    repeat_count: int
    total_cost_tokens: int
    sessions: list[RedundancySessionRef]
    source_path: str | None = None


class PlugBlockOut(BaseModel):
    """Response for ``GET /api/analytics/plug-blocks/{hash}`` (§9.2).

    ``content`` is the raw block text.  ``versions`` is populated only
    by the ``/versions`` sub-endpoint (§9.2); on the base endpoint it
    is omitted (``None``).
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    hash: str
    block_type: str
    content: str
    token_count: int
    token_count_model: str
    first_seen: int
    last_seen: int
    source_path: str | None = None


class PlugSummaryBlockOut(BaseModel):
    """One block entry inside :class:`SessionPlugSummaryOut`."""

    model_config = ConfigDict(extra="forbid")

    hash: str
    block_type: str
    tokens: int


class SessionPlugSummaryOut(BaseModel):
    """Response for ``GET /api/analytics/sessions/{id}/plug-summary`` (§9.2).

    ``status`` is derived from ``total_tokens`` vs the configured
    thresholds (``PLUG_YELLOW_THRESHOLD_TOKENS`` /
    ``PLUG_RED_THRESHOLD_TOKENS``).
    """

    model_config = ConfigDict(extra="forbid")

    total_tokens: int
    status: str = Field(
        description="'green' | 'yellow' | 'red' — derived from token count vs thresholds"
    )
    blocks: list[PlugSummaryBlockOut]

    @staticmethod
    def compute_status(total_tokens: int) -> str:
        """Return the plug status string for ``total_tokens``.

        Thresholds from ``BEARINGS_ANALYTICS_v1.md`` §8.1 via
        ``config.constants``.  Uses the module-level constants so a
        future settings-override layer can swap them without touching
        this model.
        """
        if total_tokens >= PLUG_RED_THRESHOLD_TOKENS:
            return "red"
        if total_tokens >= PLUG_YELLOW_THRESHOLD_TOKENS:
            return "yellow"
        return "green"


# ---------------------------------------------------------------------------
# Action endpoint shapes (§9.3)
# ---------------------------------------------------------------------------


class SuppressWarningIn(BaseModel):
    """Request body for ``POST /api/analytics/warnings/suppress`` (§9.3)."""

    model_config = ConfigDict(extra="forbid")

    block_hash: str = Field(min_length=64, max_length=64)
    warning_type: str = Field(description=f"One of: {_WARNING_TYPE_VALUES}")


class DraftNewSessionIn(BaseModel):
    """Request body for ``POST /api/analytics/draft-new-session`` (§9.3)."""

    model_config = ConfigDict(extra="forbid")

    source_session_id: str = Field(min_length=1)
    carry_tags: list[str] = Field(default_factory=list)


class DraftNewSessionOut(BaseModel):
    """Response for ``POST /api/analytics/draft-new-session`` (§9.3)."""

    model_config = ConfigDict(extra="forbid")

    draft_plug: str
    estimated_tokens: int
    draft_cost_tokens: dict[str, int]


class SessionFromDraftIn(BaseModel):
    """Request body for ``POST /api/analytics/sessions/from-draft`` (§9.3)."""

    model_config = ConfigDict(extra="forbid")

    draft_plug: str = Field(min_length=1)
    tags: list[str] = Field(default_factory=list)
    working_directory: str = Field(min_length=1)


class SessionFromDraftOut(BaseModel):
    """Response for ``POST /api/analytics/sessions/from-draft`` (§9.3)."""

    model_config = ConfigDict(extra="forbid")

    session_id: str


__all__ = [
    "BucketCurrentOut",
    "BucketWindowOut",
    "DraftNewSessionIn",
    "DraftNewSessionOut",
    "PlugBlockIn",
    "PlugBlockOut",
    "PlugBlocksBatchIn",
    "PlugSummaryBlockOut",
    "RedundancyBlockOut",
    "RedundancySessionRef",
    "SessionFromDraftIn",
    "SessionFromDraftOut",
    "SessionPlugSummaryOut",
    "SuppressWarningIn",
    "TagAttributionOut",
    "TurnIn",
]

File: frontend/src/lib/api/analytics.ts
/**
 * Typed client for the ``/api/analytics/`` surface
 * (``BEARINGS_ANALYTICS_v1.md`` §9).
 *
 * Backend route module: :mod:`bearings.web.routes.analytics`.
 * Pydantic shapes: :mod:`bearings.web.models.analytics`.
 *
 * Consumers:
 *
 * * :class:`InspectorAnalytics` — all read endpoints and promote actions.
 *
 * All non-2xx responses flow through :class:`ApiError`; the
 * component renders the documented error copy from
 * :data:`INSPECTOR_STRINGS.analyticsError`.
 */
import {
  ANALYTICS_ATTRIBUTION_WINDOW_WEEKLY,
  ANALYTICS_REDUNDANCY_DEFAULT_LAST_N,
  ANALYTICS_REDUNDANCY_DEFAULT_MIN_REPEATS,
  API_ANALYTICS_ATTRIBUTION_ENDPOINT,
  API_ANALYTICS_BUCKET_CURRENT_ENDPOINT,
  API_ANALYTICS_DRAFT_NEW_SESSION_ENDPOINT,
  API_ANALYTICS_PLUG_BLOCKS_BASE,
  API_ANALYTICS_REDUNDANCY_ENDPOINT,
  API_ANALYTICS_SESSION_PLUG_SUMMARY_ENDPOINT,
  API_ANALYTICS_SESSIONS_FROM_DRAFT_ENDPOINT,
  API_ANALYTICS_WARNINGS_SUPPRESS_ENDPOINT,
} from "../config";
import { getJson, postJson, type RequestOptions } from "./client";

// ---------------------------------------------------------------------------
// Response interfaces — one-to-one with Pydantic models
// ---------------------------------------------------------------------------

/**
 * One usage-window breakdown (five-hour or weekly).
 * Mirrors :class:`bearings.web.models.analytics.BucketWindowOut`.
 */
export interface BucketWindowOut {
  used: number;
  limit: number;
  /** Percentage 0-100 (rounded to 2dp by the backend). */
  percent: number;
}

/**
 * Latest ``/usage`` poll snapshot (spec §9.2 ``GET
 * /api/analytics/bucket/current``).
 * Mirrors :class:`bearings.web.models.analytics.BucketCurrentOut`.
 *
 * ``five_hour`` / ``weekly`` are ``null`` when the poller has never
 * produced data for that window.
 */
export interface BucketCurrentOut {
  five_hour: BucketWindowOut | null;
  weekly: BucketWindowOut | null;
  /** Unix ms of the snapshot's timestamp. */
  as_of: number;
}

/**
 * One row from ``GET /api/analytics/attribution`` (spec §9.2).
 * Mirrors :class:`bearings.web.models.analytics.TagAttributionOut`.
 *
 * ``tokens_by_model`` maps model id → token count.  Per spec §3.2
 * these must NOT be summed across models without normalisation; the
 * component renders them per-model to avoid mixing tokenizer output.
 */
export interface TagAttributionOut {
  tag: string;
  tokens_by_model: Record<string, number>;
  /** Fraction 0-1 of total bucket consumption. */
  share_total: number;
  /** Tokens per minute for this tag over the selected window. */
  burn_rate_per_min: number;
}

/**
 * One session reference within a redundancy block (spec §9.2).
 * Mirrors :class:`bearings.web.models.analytics.RedundancySessionRef`.
 */
export interface RedundancySessionRef {
  id: string;
  title: string;
  timestamp: number;
  tags: string[];
}

/**
 * One repeated plug block from ``GET /api/analytics/redundancy``
 * (spec §9.2 + §7.3).
 * Mirrors :class:`bearings.web.models.analytics.RedundancyBlockOut`.
 */
export interface RedundancyBlockOut {
  hash: string;
  block_type: string;
  token_count: number;
  token_count_model: string;
  repeat_count: number;
  /** ``repeat_count × token_count`` — total bucket cost of this block. */
  total_cost_tokens: number;
  source_path: string | null;
  sessions: RedundancySessionRef[];
}

/**
 * One block in a session's plug summary (spec §9.2).
 * Mirrors :class:`bearings.web.models.analytics.PlugSummaryBlockOut`.
 */
export interface PlugSummaryBlockOut {
  hash: string;
  block_type: string;
  tokens: number;
}

/**
 * Plug composition for one session (spec §9.2 ``GET
 * /api/analytics/sessions/{id}/plug-summary``).
 * Mirrors :class:`bearings.web.models.analytics.SessionPlugSummaryOut`.
 *
 * ``status`` is ``"green"`` / ``"yellow"`` / ``"red"`` per spec §8.1
 * thresholds (:data:`PLUG_YELLOW_THRESHOLD_TOKENS` /
 * :data:`PLUG_RED_THRESHOLD_TOKENS`).
 */
export interface SessionPlugSummaryOut {
  total_tokens: number;
  status: "green" | "yellow" | "red";
  blocks: PlugSummaryBlockOut[];
}

// ---------------------------------------------------------------------------
// Read clients
// ---------------------------------------------------------------------------

/**
 * Fetch the latest bucket usage snapshot.
 *
 * Returns ``{five_hour: null, weekly: null}`` when the poller has
 * never run successfully — always safe to call.
 *
 * @throws :class:`ApiError` on non-2xx (503 when DB missing).
 */
export async function getBucketCurrent(
  options: { signal?: AbortSignal } = {},
): Promise<BucketCurrentOut> {
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await getJson<BucketCurrentOut>(API_ANALYTICS_BUCKET_CURRENT_ENDPOINT, requestOptions);
}

/**
 * Fetch per-tag token attribution for the requested time window.
 *
 * ``window`` is ``"5h"`` (rolling 5-hour) or ``"weekly"`` (rolling
 * 7-day, default). Per spec §3.2 results are already grouped by model
 * in ``tokens_by_model`` — callers must not sum across models.
 *
 * @throws :class:`ApiError` on non-2xx (422 for unknown window).
 */
export async function getAttribution(
  options: {
    window?: string;
    signal?: AbortSignal;
  } = {},
): Promise<TagAttributionOut[]> {
  const window = options.window ?? ANALYTICS_ATTRIBUTION_WINDOW_WEEKLY;
  const requestOptions: RequestOptions = {
    query: [
      ["window", window],
      ["group_by", "tag"],
    ],
  };
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await getJson<TagAttributionOut[]>(API_ANALYTICS_ATTRIBUTION_ENDPOINT, requestOptions);
}

/**
 * Fetch repeated plug blocks within the requested scope.
 *
 * Results are sorted by ``total_cost_tokens`` descending (server-side).
 * ``tag`` filters to sessions with that tag name; ``null`` returns all.
 * ``lastN`` is the session-count scope (default
 * :data:`ANALYTICS_REDUNDANCY_DEFAULT_LAST_N`).
 * ``minRepeats`` is the minimum appearance count threshold (default
 * :data:`ANALYTICS_REDUNDANCY_DEFAULT_MIN_REPEATS`).
 * ``blockTypes`` is a comma-separated type filter; absent means all.
 *
 * @throws :class:`ApiError` on non-2xx (404 if tag not found, 422 for
 *   unknown block types).
 */
export async function getRedundancy(
  options: {
    tag?: string | null;
    lastN?: number;
    minRepeats?: number;
    blockTypes?: string | null;
    signal?: AbortSignal;
  } = {},
): Promise<RedundancyBlockOut[]> {
  const lastN = options.lastN ?? ANALYTICS_REDUNDANCY_DEFAULT_LAST_N;
  const minRepeats = options.minRepeats ?? ANALYTICS_REDUNDANCY_DEFAULT_MIN_REPEATS;
  const queryEntries: [string, string][] = [
    ["last_n", String(lastN)],
    ["min_repeats", String(minRepeats)],
  ];
  if (options.tag != null) {
    queryEntries.push(["tag", options.tag]);
  }
  if (options.blockTypes != null) {
    queryEntries.push(["block_types", options.blockTypes]);
  }
  const requestOptions: RequestOptions = { query: queryEntries };
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await getJson<RedundancyBlockOut[]>(API_ANALYTICS_REDUNDANCY_ENDPOINT, requestOptions);
}

/**
 * Fetch the plug composition summary for one session.
 *
 * @throws :class:`ApiError` on non-2xx (404 when session not found,
 *   503 when DB missing).
 */
export async function getSessionPlugSummary(
  sessionId: string,
  options: { signal?: AbortSignal } = {},
): Promise<SessionPlugSummaryOut> {
  const path = `${API_ANALYTICS_SESSION_PLUG_SUMMARY_ENDPOINT}/${sessionId}/plug-summary`;
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await getJson<SessionPlugSummaryOut>(path, requestOptions);
}

// ---------------------------------------------------------------------------
// Action clients
// ---------------------------------------------------------------------------

/**
 * Record that the user dismissed a plug-length warning (spec §9.3).
 *
 * Idempotent — safe to call multiple times.
 *
 * @throws :class:`ApiError` on non-2xx (422 for unknown
 *   ``warning_type``).
 */
export async function suppressWarning(
  body: { block_hash: string; warning_type: string },
  options: { signal?: AbortSignal } = {},
): Promise<{ status: string }> {
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await postJson<{ status: string }>(
    API_ANALYTICS_WARNINGS_SUPPRESS_ENDPOINT,
    body,
    requestOptions,
  );
}

// ---------------------------------------------------------------------------
// Promote action interfaces (spec §9.3)
// ---------------------------------------------------------------------------

/**
 * Request body for ``POST /api/analytics/plug-blocks/{hash}/promote-to-tag-memory``.
 * Mirrors :class:`bearings.web.models.analytics.PromoteToTagMemoryIn`.
 */
export interface PromoteToTagMemoryIn {
  tag: string;
  memory_content: string;
  auto_apply_to_next_session?: boolean;
}

/**
 * Response from ``POST /api/analytics/plug-blocks/{hash}/promote-to-tag-memory``.
 * Mirrors :class:`bearings.web.models.analytics.PromoteToTagMemoryOut`.
 */
export interface PromoteToTagMemoryOut {
  memory_id: number;
  tag: string;
}

/**
 * Request body for ``POST /api/analytics/plug-blocks/{hash}/promote-to-on-open``.
 * Mirrors :class:`bearings.web.models.analytics.PromoteToOnOpenIn`.
 */
export interface PromoteToOnOpenIn {
  working_directory: string;
  snippet: string;
}

/**
 * Response from ``POST /api/analytics/plug-blocks/{hash}/promote-to-on-open``.
 * Mirrors :class:`bearings.web.models.analytics.PromoteToOnOpenOut`.
 */
export interface PromoteToOnOpenOut {
  on_open_sh_path: string;
}

/**
 * Request body for ``POST /api/analytics/draft-new-session``.
 * Mirrors :class:`bearings.web.models.analytics.DraftNewSessionIn`.
 */
export interface DraftNewSessionIn {
  source_session_id: string;
  carry_tags?: string[];
}

/**
 * Response from ``POST /api/analytics/draft-new-session``.
 * Mirrors :class:`bearings.web.models.analytics.DraftNewSessionOut`.
 */
export interface DraftNewSessionOut {
  draft_plug: string;
  estimated_tokens: number;
  draft_cost_tokens: Record<string, number>;
}

/**
 * Request body for ``POST /api/analytics/sessions/from-draft``.
 * Mirrors :class:`bearings.web.models.analytics.SessionFromDraftIn`.
 */
export interface SessionFromDraftIn {
  draft_plug: string;
  tags?: string[];
  working_directory: string;
}

/**
 * Response from ``POST /api/analytics/sessions/from-draft``.
 * Mirrors :class:`bearings.web.models.analytics.SessionFromDraftOut`.
 */
export interface SessionFromDraftOut {
  session_id: string;
}

// ---------------------------------------------------------------------------
// Promote action client functions (spec §9.3)
// ---------------------------------------------------------------------------

/**
 * Promote a plug block to a tag memory (spec §9.3 §7.5).
 *
 * Idempotent — re-promoting the same block to the same tag returns the
 * existing memory without creating a duplicate.
 *
 * @throws :class:`ApiError` on non-2xx (404 when block or tag not found).
 */
export async function promoteToTagMemory(
  hash: string,
  body: PromoteToTagMemoryIn,
  options: { signal?: AbortSignal } = {},
): Promise<PromoteToTagMemoryOut> {
  const endpoint = `${API_ANALYTICS_PLUG_BLOCKS_BASE}/${hash}/promote-to-tag-memory`;
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await postJson<PromoteToTagMemoryOut>(endpoint, body, requestOptions);
}

/**
 * Promote a plug block to ``<working_directory>/.bearings/on_open.sh`` (spec §9.3 §7.5).
 *
 * Idempotent — re-promoting the same snippet does not duplicate it.
 *
 * @throws :class:`ApiError` on non-2xx (404 when block not found,
 *   422 when working_directory is not an existing directory).
 */
export async function promoteToOnOpen(
  hash: string,
  body: PromoteToOnOpenIn,
  options: { signal?: AbortSignal } = {},
): Promise<PromoteToOnOpenOut> {
  const endpoint = `${API_ANALYTICS_PLUG_BLOCKS_BASE}/${hash}/promote-to-on-open`;
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await postJson<PromoteToOnOpenOut>(endpoint, body, requestOptions);
}

/**
 * Generate a draft plug for a new session continuing from a source session
 * (spec §9.3 §8.4).
 *
 * @throws :class:`ApiError` on non-2xx (404 when source session not found).
 */
export async function draftNewSession(
  body: DraftNewSessionIn,
  options: { signal?: AbortSignal } = {},
): Promise<DraftNewSessionOut> {
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await postJson<DraftNewSessionOut>(
    API_ANALYTICS_DRAFT_NEW_SESSION_ENDPOINT,
    body,
    requestOptions,
  );
}

/**
 * Create a new session from a user-reviewed plug draft (spec §9.3 §8.4).
 *
 * @throws :class:`ApiError` on non-2xx (404 when any tag not found).
 */
export async function createSessionFromDraft(
  body: SessionFromDraftIn,
  options: { signal?: AbortSignal } = {},
): Promise<SessionFromDraftOut> {
  const requestOptions: RequestOptions = {};
  if (options.signal !== undefined) {
    requestOptions.signal = options.signal;
  }
  return await postJson<SessionFromDraftOut>(
    API_ANALYTICS_SESSIONS_FROM_DRAFT_ENDPOINT,
    body,
    requestOptions,
  );
}
